### PR TITLE
Removing matplotlib from the mandatory requirements

### PR DESCRIPTION
**Pull request by alexsu52**
### Changes

matplotlib dependency has been moved under dev extra.

### Reason for changes

OpenVINO notebooks have compatibility issues with matplotlib dependency.

### Related tickets

Custom request

### Tests

N/A

### CODE REVIEW

File: nncf/experimental/torch/nas/bootstrapNAS/search/search.py [hunk 1]
```diff
@@ -50,6 +48,16 @@
 from nncf.experimental.torch.nas.bootstrapNAS.elasticity.multi_elasticity_handler import SubnetConfig
 from nncf.torch.nncf_network import NNCFNetwork
 
+try:
+    import matplotlib.pyplot as plt
```
Comment (ljaljushkin) [inline nncf/experimental/torch/nas/bootstrapNAS/search/search.py:52]: To be more consistent, it makes sense to do the same for hawq_debug.py file
could you do that in the PR?

Comment (alexsu52) [reply nncf/experimental/torch/nas/bootstrapNAS/search/search.py]: Honestly, I thought that the debug code should fail if the dependencies are not meet But nevertheless, I did it.

File: nncf/common/utils/decorators.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+from importlib import import_module
+from typing import Callable, List
+
+from nncf.common.logging import nncf_logger
+
+IMPORTED_DEPENDENCIES = {}
+
+
+def bypass_noreturn_function(dependencies: List[str]) -> Callable:
```
Comment (vshampor) [inline nncf/common/utils/decorators.py:22]: `skip_if_dependency_unavailable`

`typing.NoReturn` that you are referring to here is actually a wrong type hint - it marks functions that *do not return* (e.g. functions that always raise an exception), not functions that return `None` explicitly or implicitly.

File: nncf/common/utils/decorators.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    """
+    Decorator to bypass a noreturn function if dependencies are not met.
+
+    :param func: A noreturn function to decorate
+    :return: Decorated function
+    """
+    def wrap(func: Callable) -> Callable:
```
Comment (vshampor) [inline nncf/common/utils/decorators.py:29]: `Callable[..., None]` to mark that the callable to be wrapped should only be a function that does not return a specific return value other than `None`

File: nncf/common/utils/decorators.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+from nncf.common.logging import nncf_logger
+
+IMPORTED_DEPENDENCIES = {}
+
+
+def bypass_noreturn_function(dependencies: List[str]) -> Callable:
+    """
+    Decorator to bypass a noreturn function if dependencies are not met.
+
+    :param func: A noreturn function to decorate
```
Comment (vshampor) [inline nncf/common/utils/decorators.py:26]: Should probably describe the `dependencies` parameter instead.

**Review by vshampor — CHANGES_REQUESTED** [review]

File: nncf/common/utils/decorators.py (hunk 2 above)
Comment (alexsu52) [reply nncf/common/utils/decorators.py]: done

File: nncf/common/utils/decorators.py (hunk 3 above)
Comment (alexsu52) [reply nncf/common/utils/decorators.py]: done

File: nncf/common/utils/decorators.py (hunk 4 above)
Comment (alexsu52) [reply nncf/common/utils/decorators.py]: done

**Review by vshampor — APPROVED** [review]

**Pull request merged**